### PR TITLE
core: implement `fuzzy` searching and highlights

### PR DESCRIPTION
**Pull request by vince-fugnitto**
#### What it does

Fixes: https://github.com/eclipse-theia/theia/issues/9851

The pull-request includes updates to the following methods to make them support `fuzzy` searching and highlighting:
- `QuickInputService.filterItems`
- `QuickInputService.findMatches`

Previously, both methods only ever supported strict `substring` searches and highlights which resulted in many items not being displayed properly after https://github.com/eclipse-theia/theia/pull/9702. The change now makes both methods `fuzzy` aware, and will match items on all their properties which are present in the UI (`label`, `description`, `detail`). The `findMatches` method is also enhanced to properly compute highlights, even when we match `fuzzy`-ingly. 

The pull-request also introduces appropriate unit-tests in order to properly test both methods under various conditions and use-cases.

https://user-images.githubusercontent.com/40359487/129908954-fe9286e3-57e4-4b20-b541-cda1affb6c5a.mov

#### How to test

- the new test-cases should successfully pass.
- it should now be possible to use the `quick-command` palette (<kbd>F1</kbd>) and perform searches will result in fuzzy matches, and it will display more results than `master`.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: vince-fugnitto <vincent.fugnitto@ericsson.com>

### CODE REVIEW

File: packages/core/src/browser/quick-input/quick-input-service.ts
Comment (msujew) [inline packages/core/src/browser/quick-input/quick-input-service.ts:244]: These highlightings are exclusive, so highlights for `{ label: 'test', description: 'test' }` when searching for `test` will only highlight the label. Compare this with the behavior of vscode, where both are highlighted:

grafik

File: packages/core/src/browser/quick-input/quick-input-service.ts
Comment (msujew) [inline packages/core/src/browser/quick-input/quick-input-service.ts:247]: Nitpick: This seems a bit redundant.

File: packages/core/src/browser/quick-input/quick-input-service.spec.ts
Comment (colin-grant-work) [inline packages/core/src/browser/quick-input/quick-input-service.spec.ts:29]: 'ac' might be a more interesting third case, rather than another one that matches consecutively?

**Comment by msujew** [comment]
While trying to test the behavior of labels & descriptions on items, I noticed some weird behavior in the fuzzy search when using the `Task: run task...` command in the Theia repo. 

grafik

Most of the `postinstall` label should be matched by the `posinstall` search value, but it's not matched/highlighted at all.

This however works when searching for commands:

grafik

**Comment by vince-fugnitto** [comment]
> While trying to test the behavior of labels & descriptions on items, I noticed some weird behavior in the fuzzy search when using the `Task: run task...` command in the Theia repo.

@msujew the prefix `task ` quick-open menu does in fact display the proper filtering and highlighting. It looks like `run task...` does not use the `quick-input-service` in order to display results but rather the `quick-pick`.

**Comment by msujew** [comment]
@vince-fugnitto Alright, I see. I guess that's to be solved in a separate PR, so feel free to ignore my comment 👍

**Review by msujew — APPROVED** [review]
Can confirm that the fuzzy search works well. I can filter and highlight both on label and other properties:
grafik

The code and tests look good to me as well.

This also resolves an issue I had in my i18n PR, since users couldn't search for the original name of commands :)

**Comment by tsmaeder** [comment]
@vince-fugnitto anything keeping this from being merged?

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto anything keeping this from being merged?

@tsmaeder I was waiting for a second reviewer, but I can merge early next week since it's Friday :smile:

**Review by colin-grant-work — APPROVED** [review]
Working well for me 👍

**Pull request merged**
